### PR TITLE
[CA-1242] Sam Admin APIs

### PR DESCRIPTION
**Pull request by marctalbott**
Ticket: <Link to Jira ticket> https://broadworkbench.atlassian.net/browse/CA-1242
<Put notes here to help reviewer understand this PR>

---

**PR checklist**

- [ ] I've followed [the instructions](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've filled out the [Security Risk Assessment](https://broadinstitute.github.io/dsp-appsec-security-risk-assessment/) and attached the result to the JIRA ticket

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/ResourceRoutes.scala [hunk 1]
```diff
@@ -36,6 +36,67 @@ trait ResourceRoutes extends UserInfoDirectives with SecurityDirectives with Sam
       case None => throw new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.NotFound, s"resource type ${name.value} not found"))
     }
 
+  def adminResourceRotues: server.Route =
```
Comment (s-rubenstein) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/api/ResourceRoutes.scala]: TOL: I had actually been thinking this should maybe be in a new file, since the routes aren't necessarily related (although there is a lot of shared functionality)

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/CloudExtensions.scala [hunk 2]
```diff
@@ -66,6 +68,8 @@ trait CloudExtensionsInitializer {
 trait NoExtensions extends CloudExtensions {
   override def isWorkbenchAdmin(memberEmail: WorkbenchEmail): Future[Boolean] = Future.successful(true)
 
+  override def isSamSuperAdmin(memberEmail: WorkbenchEmail): Future[Boolean] = Future.successful(true)
```
Comment (s-rubenstein) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/service/CloudExtensions.scala:71]: Is there a todo to implement this function? Or if not, how does this work?

**Review by s-rubenstein — COMMENTED** [review]
What is the plan to add the swagger info here? Are you doing it as a follow up, or do you want to do that in this PR?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/ResourceRoutes.scala (hunk 1 above)
Comment (marctalbott) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/api/ResourceRoutes.scala]: Good call, I followed the admin routes pattern we used for admin user routes, but I think it's separate enough to pull into a different file

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/CloudExtensions.scala (hunk 2 above)
Comment (marctalbott) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/service/CloudExtensions.scala:71]: `NoExtensions` is for running Sam without connecting to Google which we only do in tests. I'm not sure _why_ we have it defined here instead of as a test file or something like that, but that's why it's just stubbed out like this

Comment (s-rubenstein) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/service/CloudExtensions.scala:71]: Yeah, I was suspecting it was for tests, but the fact that it wasn't in a spec folder had me second guessing myself. Sounds good.

**Comment by marctalbott** [comment]
> What is the plan to add the swagger info here? Are you doing it as a follow up, or do you want to do that in this PR?

I'll add it to this PR, otherwise I think I'll forget

**Review by s-rubenstein — APPROVED** [review]
LGTM

**Comment by marctalbott** [comment]
This turned into the feature branch at some point, but there's not really any reason for that afaict so I'm merging it into the `jamboree` branch for simplicity's sake

EDIT: scratch that, I can only squash and merge in the Sam repo and I don't want to lose the commit history. I'll be figuring something else out

**Comment by andy7i** [comment]
jenkins retest

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala [hunk 3]
```diff
@@ -221,7 +223,7 @@ class ResourceService(
     })
 
   @VisibleForTesting
-  private[service] def createPolicy(policyIdentity: FullyQualifiedPolicyId, members: Set[WorkbenchSubject], roles: Set[ResourceRoleName], actions: Set[ResourceAction], descendantPermissions: Set[AccessPolicyDescendantPermissions], samRequestContext: SamRequestContext): IO[AccessPolicy] =
+  def createPolicy(policyIdentity: FullyQualifiedPolicyId, members: Set[WorkbenchSubject], roles: Set[ResourceRoleName], actions: Set[ResourceAction], descendantPermissions: Set[AccessPolicyDescendantPermissions], samRequestContext: SamRequestContext): IO[AccessPolicy] =
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala:226]: why this change?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala [hunk 4]
```diff
@@ -312,8 +315,29 @@ class ResourceService(
...
   }
 
+  /**
+    * Overwrites an existing admin policy, saves a new one if it doesn't exist yet.  */
+  def overwriteAdminPolicy(resourceType: ResourceType, policyName: AccessPolicyName, resource: FullyQualifiedResourceId, policyMembership: AccessPolicyMembership, samRequestContext: SamRequestContext): IO[AccessPolicy] =
+    failUnlessAllAdminEmailDomainsWhitelisted(policyMembership) *>
+      overwritePolicy(resourceType, policyName, resource, policyMembership, samRequestContext)
+
+  /** Performs additional verification that members are only users in the correct email domain */
+  def failUnlessAllAdminEmailDomainsWhitelisted(membership: AccessPolicyMembership): IO[Unit] =
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala]:
```suggestion
  def failUnlessAllAdminEmailDomainsAllowed(membership: AccessPolicyMembership): IO[Unit] =
```

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/config/AppConfig.scala [hunk 5]
```diff
@@ -148,25 +151,37 @@ object AppConfig {
     LiquibaseConfig(config.getString("changelog"), config.getBoolean("initWithLiquibase"))
   }
 
+  final case class AdminConfig(superAdminsGroup: WorkbenchEmail, adminEmailDomains: Set[String])
+
+  implicit val adminConfigReader: ValueReader[AdminConfig] = ValueReader.relative { config =>
+    AdminConfig(
+      superAdminsGroup = WorkbenchEmail(config.getString("superAdminsGroup")),
+      adminEmailDomains = config.as[Set[String]]("emailDomains")
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/config/AppConfig.scala]: I would call this `allowedAdminEmailDomains` in the config file

**Review by dvoet — APPROVED** [review]
a couple minor tweaks around naming

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala (hunk 3 above)
Comment (ehigham) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala:226]: Needs to be public for the new admin resource tests

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala (hunk 4 above)
Comment (ehigham) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala]: Thank you very much - I was struggling to come up with a different verb.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/config/AppConfig.scala (hunk 5 above)
Comment (ehigham) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/config/AppConfig.scala]: Good idea
